// react-states/devtools/Resizer.js
import * as React from 'react';
import { createStates, createActions, transition, useTransition, match, useEnter, } from '../';
import { colors } from './styles';
const states = createStates({
    IDLE: () => ({}),
    DETECTING_RESIZE: (initialX) => ({ initialX }),
    RESIZING: (x) => ({ x }),
});
const actions = createActions({
    onMouseMove: (x) => ({ x }),
    onMouseUp: (x) => ({ x }),
    onMouseUpResizer: () => ({}),
    onMouseDown: (x) => ({ x }),
});
const reducer = (state, action) => transition(state, action, {
    IDLE: {
        onMouseDown: (_, { x }) => states.DETECTING_RESIZE(x),
    },
    DETECTING_RESIZE: {
        onMouseMove: (state, { x }) => {
            if (Math.abs(x - state.initialX) > 3) {
                return states.RESIZING(x);
            }
            return state;
        },
        onMouseUp: () => states.IDLE(),
        onMouseUpResizer: () => states.IDLE(),
    },
    RESIZING: {
        onMouseMove: (_, { x }) => states.RESIZING(x),
        onMouseUp: () => states.IDLE(),
    },
});
export const Resizer = ({ onResize, onClick, isOpen, }) => {
    const [resizer, dispatch] = React.useReducer(reducer, {
        state: 'IDLE',
    });
    const { onMouseDown, onMouseMove, onMouseUp, onMouseUpResizer } = actions(dispatch);
    useEnter(resizer, ['DETECTING_RESIZE', 'RESIZING'], () => {
        const onMouseMoveListener = (event) => onMouseMove(event.clientX);
        const onMouseUpListener = (event) => onMouseUp(event.clientX);
        window.addEventListener('mousemove', onMouseMoveListener);
        window.addEventListener('mouseup', onMouseUpListener);
        return () => {
            window.removeEventListener('mousemove', onMouseMoveListener);
            window.removeEventListener('mouseup', onMouseUpListener);
        };
    });
    useTransition(resizer, ['DETECTING_RESIZE => onMouseMove => DETECTING_RESIZE', 'RESIZING => onMouseMove => RESIZING'], (_, { x }) => {
        onResize(window.innerWidth - x);
    });
    useTransition(resizer, 'DETECTING_RESIZE => onMouseUpResizer => IDLE', () => onClick());
    const style = {
        position: 'absolute',
        height: '100%',
        width: '10px',
        backgroundColor: colors.blue,
        userSelect: 'none',
        zIndex: 99999999,
    };
    return match(resizer, {
        IDLE: () => React.createElement("div", { style: style, onMouseDown: (event) => onMouseDown(event.clientX) }),
        DETECTING_RESIZE: () => React.createElement("div", { style: style, onMouseUp: () => onMouseUpResizer() }),
        RESIZING: () => React.createElement("div", { style: style }),
    });
};
