// react-states/cjs/devtools/History.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.History = void 0;
const tslib_1 = require("tslib");
const React = tslib_1.__importStar(require("react"));
const styles_1 = require("./styles");
const ValueInspector_1 = tslib_1.__importDefault(require("./ValueInspector"));
exports.History = React.memo(({ history, filterIgnored }) => {
    return (React.createElement("ul", { style: {
            listStyleType: 'none',
            padding: 0,
            color: styles_1.colors.text,
            marginBottom: '0.5rem',
            fontSize: '14px',
        } }, history.map((item, index) => {
        if (item.type === 'state') {
            return (React.createElement("li", { key: index, style: {
                    display: 'flex',
                    alignItems: 'center',
                    marginBottom: '0.25rem',
                } },
                React.createElement("span", { style: {
                        marginRight: '0.25rem',
                        fontSize: '12px',
                        color: styles_1.colors.orange,
                    } }, "state"),
                React.createElement("span", { style: {
                        marginRight: '0.25rem',
                    } }, item.state.state),
                React.createElement(ValueInspector_1.default, { value: item.state, small: true })));
        }
        if (item.ignored && filterIgnored) {
            return null;
        }
        return (React.createElement("li", { key: index, style: {
                display: 'flex',
                alignItems: 'center',
                marginBottom: '0.25rem',
                opacity: item.ignored ? 0.5 : 1,
            } },
            React.createElement("span", { style: {
                    marginRight: '0.25rem',
                    fontSize: '12px',
                    color: styles_1.colors.green,
                } }, "action"),
            React.createElement("span", { style: { marginRight: '0.25rem' } }, item.action.type),
            React.createElement(ValueInspector_1.default, { value: item.action, small: true })));
    })));
});
