// react-states/cjs/devtools/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ok = exports.newState = exports.editValuePopup = exports.editValueWrapper = exports.keyCount = exports.nestedChildren = exports.genericValue = exports.inlineClass = exports.otherValue = exports.stringValue = exports.bracket = exports.inlineNested = exports.toolIcon = exports.key = exports.smallWrapper = exports.wrapper = exports.colors = void 0;
exports.colors = {
    purple: '#a855f7',
    yellow: '#eab308',
    green: '#5DCC67',
    blue: '#3b82f6',
    orange: '#F7A239',
    red: '#FF3B30',
    foreground: 'rgb(36, 36, 36)',
    background: 'rgb(21, 21, 21)',
    border: 'rgb(36, 36, 36)',
    text: 'rgb(153, 153, 153)',
    highlight: 'hsl(0, 0%, 85%)',
};
exports.wrapper = {
    fontFamily: "'Source Code Pro', monospace",
    fontSize: 16,
    lineHeight: '24px',
    color: exports.colors.highlight,
};
exports.smallWrapper = {
    fontFamily: "'Source Code Pro', monospace",
    fontSize: 12,
    lineHeight: '16px',
};
exports.key = {
    marginRight: 5,
    color: exports.colors.text,
    cursor: 'pointer',
    /*
      ":hover": {
        opacity: 0.75
      }
      */
};
exports.toolIcon = {
    margin: '0 0.75rem',
};
exports.inlineNested = {
    display: 'flex',
    alignItems: 'center',
    cursor: 'pointer',
};
const bracket = (pointer) => ({
    display: 'flex',
    alignItems: 'center',
    cursor: pointer ? 'pointer' : 'default',
});
exports.bracket = bracket;
exports.stringValue = {
    display: 'flex',
    alignItems: 'center',
    color: exports.colors.orange,
};
exports.otherValue = {
    display: 'flex',
    alignItems: 'center',
    color: exports.colors.purple,
};
exports.inlineClass = {
    color: exports.colors.purple,
    marginRight: '0.5rem',
};
exports.genericValue = {
    display: 'flex',
    alignItems: 'center',
    color: exports.colors.blue,
};
exports.nestedChildren = {
    paddingLeft: '1rem',
};
exports.keyCount = {
    fontsize: 14,
    color: exports.colors.highlight,
};
exports.editValueWrapper = {
    position: 'relative',
};
exports.editValuePopup = {
    position: 'absolute',
    width: 400,
    height: 100,
    top: 0,
    left: 0,
    boxShadow: '0px 10px 13px 0px rgba(0,0,0,0.1)',
};
exports.newState = {
    fontFamily: 'inherit',
    fontSize: 16,
    border: '2px solid transparent',
    backgroundColor: exports.colors.text,
    color: exports.colors.foreground,
    outline: 'none',
    borderRadius: 3,
    width: '100%',
    height: '100%',
    boxSizing: 'border-box',
};
exports.ok = {
    position: 'absolute',
    cursor: 'pointer',
    top: 0,
    right: 0,
    fontSize: 10,
    border: 0,
    outline: 'none',
    padding: '0.25rem 0.5rem',
    opacity: 0.5,
    color: exports.colors.background,
};
